### DISCUSSION TITLE
[Plugin Request] Wrap fields in StoreObject for Apollo Client

### DISCUSSION THREAD

**Discussion by bennypowers** [dotansimha/graphql-code-generator#4730] [General]
**Is your feature request related to a problem? Please describe.**

Hey Hey. Many new apollo client 3 methods expect an instance of `StoreObject`, which is exported from `@apollo/client/cache`. The base type is` __typename`, along with an **index signature** to `StoreProperty`.

**Describe the solution you'd like**
It would be cool if graphql-codegen would take an option to wrap types in `StoreObject`

**Describe alternatives you've considered**
from chat: 
dotansimha
> **dotansimha**: So basically, since the StoreObject is just a type with __typename set, it means that you can pass any type that is generated from GraphQL Codegen (as long as you have addTypename: string, nonOptionalTypename: true set in your config). 
> Since TypeScript isn't a nominal language (at-least most of it..), you can pass an object that is similar in structure, and it will pass validation. In that case, you just need __typename to be non-optional, and you can achieve that by setting the correct configuration.
> 
> **bennyp**: dies because of no index signature, though

**Additional context**
from chat:
> **dotansimha**: It might be the storeFieldName they have there. I guess it could be simpler with Record ? 
Can you please report this as issue in codegen? I'll look into that :slight_smile:
I'm not sure that adding an index signature is the right way to go, since it will break type-safety in some cases. Maybe we can try to wrap it with StoreObject as you suggested, but ideally, maybe AC3 should have better typings in that case?
> 
> **bennyp**: yeah it's a bit upsetting. I guess they wanted to express that "all the things need to be store props"
probably something to do with the new _refs system?

**Comment by dotansimha** [comment]
Thank you for this great suggestion, @bennypowers . I'm moving it to Discussions (we track plugins requests there)

**Reply by bennypowers** [reply]
updated to latest versions of everything to day and this seems to be resolved

**Reply by bennypowers** [reply]
Ignore that last comment, tsc was acting up, this feature/bug is still relevant
